File: Comportamentais/ChainOfResponsibility.java
package Comportamentais;

// Definindo uma classe abstrata para manipuladores
abstract class Handler {
    protected Handler successor;

    public void setSuccessor(Handler successor) {
        this.successor = successor;
    }

    public abstract void handleRequest(int request);
}

// Implementando manipuladores específicos
class LowLevelHandler extends Handler {
    @Override
    public void handleRequest(int request) {
        if (request < 10) {
            System.out.println("LowLevelHandler: Processando o pedido " + request);
        } else {
            System.out.println("LowLevelHandler: Passando para o próximo.");
            if (successor != null) {
                successor.handleRequest(request);
            }
        }
    }
}

class HighLevelHandler extends Handler {
    @Override
    public void handleRequest(int request) {
        if (request >= 10) {
            System.out.println("HighLevelHandler: Processando o pedido " + request);
        } else {
            System.out.println("HighLevelHandler: Não pode processar.");
        }
    }
}

// Configuração da cadeia de responsabilidade
public class ChainOfResponsibility {
    public static void main(String[] args) {
        Handler lowLevelHandler = new LowLevelHandler();
        Handler highLevelHandler = new HighLevelHandler();
        lowLevelHandler.setSuccessor(highLevelHandler);

        lowLevelHandler.handleRequest(5);  // Processado pelo LowLevelHandler
        lowLevelHandler.handleRequest(20); // Processado pelo HighLevelHandler
    }
}

File: Comportamentais/ChainOfResponsibility.py
# Definindo uma classe base para manipuladores
class Handler:
    def __init__(self, successor=None):
        self.successor = successor  # Próximo manipulador na cadeia

    def handle_request(self, request):
        if self.successor:
            self.successor.handle_request(request)

# Implementando manipuladores específicos
class LowLevelHandler(Handler):
    def handle_request(self, request):
        if request < 10:
            print(f"LowLevelHandler: Processando o pedido {request}")
        else:
            print("LowLevelHandler: Passando para o próximo.")
            super().handle_request(request)

class HighLevelHandler(Handler):
    def handle_request(self, request):
        if request >= 10:
            print(f"HighLevelHandler: Processando o pedido {request}")
        else:
            print("HighLevelHandler: Não pode processar.")

# Configuração da cadeia de responsabilidade
handler = LowLevelHandler(HighLevelHandler())
handler.handle_request(5)  # Espera-se que seja processado pelo LowLevelHandler
handler.handle_request(20)  # Espera-se que seja processado pelo HighLevelHandler
